compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -186,18 +187,32 @@ pub fn unsized_info<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
                 old_info
             }
         }
-        (_, &ty::Dynamic(ref data, ..)) => {
-            let vtable_ptr_ty = cx.scalar_pair_element_backend_type(
-                cx.layout_of(cx.tcx().mk_mut_ptr(target)),
-                1,
-                true,
-            );
+        (_, &ty::Dynamic(ref data, _, target_dyn_kind)) => {
+            let vtable_ptr_ty = vtable_ptr_ty(cx, target, target_dyn_kind);
             cx.const_ptrcast(meth::get_vtable(cx, source, data.principal()), vtable_ptr_ty)
         }
         _ => bug!("unsized_info: invalid unsizing {:?} -> {:?}", source, target),
     }
 }
 
+// Returns the vtable pointer type of a `dyn` or `dyn*` type
+fn vtable_ptr_ty<'tcx, Cx: CodegenMethods<'tcx>>(
+    cx: &Cx,
+    target: Ty<'tcx>,
+    kind: ty::DynKind,
+) -> <Cx as BackendTypes>::Type {
+    cx.scalar_pair_element_backend_type(
+        cx.layout_of(match kind {
+            // vtable is the second field of `*mut dyn Trait`
+            ty::Dyn => cx.tcx().mk_mut_ptr(target),
+            // vtable is the second field of `dyn* Trait`
+            ty::DynStar => target,
```

Comment:
It's a little surprising to me that these are different. Is that just an artifact of how `dyn*` is represented, or is there a reason they can't be the same?

Oh, I guess for `ty::Dyn`, `target` is an unsized `dyn Trait` object, so here we convert it into a pointer, while `ty::DynStar` is already a pointer so we can just use `target`.

Reply:
`&mut dyn*` is `usize`-sized, so getting the 1st field on that would fail. `dyn` is unsized, so getting the 1st field on that would fail. The vtable is `(&mut dyn Trait)[1]` or `(dyn* Trait)[1]` basically.